cloudevents/event_v02.go

```diff
@@ -0,0 +1,261 @@
+		Extensions:         make(map[string]interface{}),
+	}
+	for k, v := range ec.Extensions {
+		// eventTypeVersion was retired in v0.2
+		if strings.EqualFold(k, "eventTypeVersion") {
+			etv, ok := v.(string)
+			if ok {
+				ret.EventTypeVersion = etv
+			}
+			continue
+		}
+		ret.Extensions[k] = v
+	}
+	return ret
+}
+
+// AsV02 implements the ContextTranslator interface.
+func (ec V02EventContext) AsV02() V02EventContext {
+	return ec
+}
+
+// AsHeaders implements the BinarySender interface.
+func (ec V02EventContext) AsHeaders() http.Header {
+	h := http.Header{}
+	h.Set("CE-"+fieldSpecVersion, ec.SpecVersion)
+	h.Set("CE-"+fieldType, ec.Type)
+	h.Set("CE-"+fieldSource, ec.Source)
+	h.Set("CE-"+fieldID, ec.ID)
+	if ec.SpecVersion == "" {
+		h.Set("CE-"+fieldSpecVersion, V02CloudEventsVersion)
+	}
+	if !ec.Time.IsZero() {
+		h.Set("CE-"+fieldTime, ec.Time.Format(time.RFC3339Nano))
+	}
+	if ec.SchemaURL != "" {
+		h.Set("CE-"+fieldSchemaURL, ec.SchemaURL)
+	}
+	if ec.ContentType != "" {
+		h.Set(headerContentType, ec.ContentType)
+	}
+	for k, v := range ec.Extensions {
+		// Per spec, map-valued extensions are converted to a list of headers as:
+		// CE-attrib-key
+		if mapVal, ok := v.(map[string]interface{}); ok {
+			for subkey, subval := range mapVal {
+				data := fmt.Sprint(subval)
+				if encoded, err := json.Marshal(subval); err == nil {
+					data = string(encoded)
+				}
+				h.Set("CE-"+k+"-"+subkey, data)
+			}
+			continue
+		}
+		data := fmt.Sprint(v)
+		if encoded, err := json.Marshal(v); err == nil {
+			data = string(encoded)
+		}
+		h.Set("CE-"+k, data)
+	}
+
+	return h
+}
+
+// FromHeaders implements the BinaryLoader interface.
+func (ec *V02EventContext) FromHeaders(in http.Header) error {
+	missingField := func(name string) error {
+		if in.Get("CE-"+name) == "" {
+			return fmt.Errorf("Missing field %q in %v: %q", "CE-"+name, in, in.Get("CE-"+name))
+		}
+		return nil
+	}
+	err := anyError(
+		missingField(fieldSpecVersion),
+		missingField(fieldID),
+		missingField(fieldType),
+		missingField(fieldSource),
+	)
+	if err != nil {
+		return err
+	}
+	data := V02EventContext{
+		ContentType: in.Get(headerContentType),
+		Extensions:  make(map[string]interface{}),
+	}
+	// Extensions and top-level fields are mixed under "CE-" headers.
+	// Extract them all here rather than trying to clear fields in headers.
+	for k, v := range in {
+		if strings.EqualFold(k[:len("CE-")], "CE-") {
+			key, value := strings.ToLower(string(k[len("CE-"):])), v[0]
+			switch key {
+			case fieldSpecVersion:
+				data.SpecVersion = value
+			case fieldType:
+				data.Type = value
+			case fieldSource:
+				data.Source = value
+			case fieldID:
+				data.ID = value
+			case fieldSchemaURL:
+				data.SchemaURL = value
+			case fieldTime:
+				if data.Time, err = time.Parse(time.RFC3339Nano, value); err != nil {
+					return err
+				}
+			default:
+				var tmp interface{}
+				if err = json.Unmarshal([]byte(value), &tmp); err != nil {
```

Comment:
Do we care about the error here, or we expect them to happen?

Reply:
The error will happen if you have an unquoted string, like `CE-SpecVersion: v0.2`.

Right now, it seems more helpful to interpret these as strings, rather than reject them.

Reply:
I added some tests adjacent to this, but I'm not sure we have something which sends a raw HTTP request rather than something created by a marshaller.

Mind if I add the raw request testing in another PR (in part because the raw HTTP behavior is not well-defined in the CloudEvents 0.2 spec in terms of error handling)?